ml_swissknife/db_utils.py

```diff
@@ -1,231 +1,210 @@
+            f"There are no tables you referred to in your SQL query, we cannot enforce column types in dataframe."
+        )
 
-    if is_enforce_numeric:
-        cols_to_numeric = [r["name"] for _, r in table_info.iterrows() if r["type"] in ["INTEGER", "FLOAT"]]
-        for c in cols_to_numeric:
-            if c in df.columns:
-                df[c] = df[c].apply(pd.to_numeric, errors="coerce")
+    if table_name_to_enforce_types is not None:
+        db_table = sa.Table(table_name_to_enforce_types, sa.MetaData(), autoload_with=engine)
+        enforce_type_cols_df_(df, db_table)
 
     return df
 
 
-def get_values_from_keys(database, table, df, is_rm_duplicates=False):
-    """Given a dataframe containing the primary keys of a table, will return the corresponding rows"""
-    keys = ", ".join(df.columns)
-    list_of_tuples = list(zip(*[df[c] for c in df.columns]))
+def delete_rows_from_db(database: Union[str, sa.engine.base.Engine], table_name: str, df: pd.DataFrame):
+    """Delete rows from a table in a SQLite database based on the values of a dataframe.
 
-    list_of_placeholders = [tuple("?" for _ in range(len(df.columns)))] * len(list_of_tuples)
-    placeholders = str(list_of_placeholders)[1:-1].replace("'?'", "?")
+    Parameters
+    ----------
+    database : str or engine
+        The URL to the database to connect to, or the sqlalchemy engine.
 
-    flattened_values = [item for sublist in list_of_tuples for item in sublist]
+    table_name : str
+        The name of the table to delete rows from.
+
+    df : pd.DataFrame
+        The dataframe containing the rows to delete.
+
+    Examples
+    --------
+    >>> df = sql_to_df(sql="SELECT * FROM likert_annotations LIMIT 3", database=database)
+    >>> len(df)
+    3
+    >>> delete_rows_from_db(database=database,  df=df[["input_id", "output", "annotator"]], table_name="likert_annotations")
+    >>> len(get_values_from_keys(database=database,  df=df[["input_id", "output", "annotator"]], table_name="likert_annotations"))
+    0
+    """
+    engine = get_engine(database)
+    db_table = sa.Table(table_name, sa.MetaData(), autoload_with=engine)
 
-    try:
-        # this is much faster an memory efficient but sometimes has some formatting issues
-        # Example sql query is built as follows:
-        #   SELECT * FROM likert_annotations WHERE (input_id, output) IN (VALUES (?, ?), (?, ?), (?, ?))
-        out = sql_to_df(
-            database=database,
-            sql=f"""SELECT * FROM {table} WHERE ({keys}) IN (VALUES {placeholders})""",
-            params=flattened_values,
-        )
-    except:
-        # this reads everything in memory which is less efficient but more robust
-        out = sql_to_df(
-            database=database,
-            sql=f"SELECT * FROM {table}",
-        )
+    sql_where = get_sql_where_from_df(engine, table=db_table, df=df)
+    delete = db_table.delete().where(sql_where)
+    execute_sql(database=engine, sql=delete)
```

Comment:
this seems to be the only usage of `execute_sql`. do we need to have the helper `execute_sql`?

Reply:
in human feedback it's used